drivers/hmis/app/models/hmis/hud/enrollment.rb

```diff
@@ -287,7 +287,22 @@ def self.generate_enrollment_id
     generate_uuid
   end
 
-  def save_in_progress
+  def save_new_enrollment!
+    raise 'Unexpected: save_new_enrollment called on a persisted enrollment' if persisted?
+
+    if Hmis::ProjectAutoEnterConfig.detect_best_config_for_project(project)
+      # If auto-enter is configured for this project, save as non-WIP and generate an empty intake.
+      save_not_in_progress!
+      build_synthetic_intake_assessment.save!
```

Comment:
maybe these should be in a transaction together?
```suggestion
      transaction do 
        save_not_in_progress!
        build_synthetic_intake_assessment.save!
      end
```

Reply:
`save_new_enrollment!` is expected to be called within a transaction already. Also I just noticed `save_not_in_progress!` has a transaction in it, so we already have a problem with nesting.

From my understanding, we definitely DON'T want nested transactions within a mutation execution, because that would lead to rollbacks being ignored. Even adding `requires_new`, though, isn't what we want in these cases. I'm looking at https://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html#module-ActiveRecord::Transactions::ClassMethods-label-Nested+transactions.

Actually this made me realize we have a few places with nested transactions that probably are at risk of committing data even if a rollback is raised. Specifically: [here](https://github.com/greenriver/hmis-warehouse/blob/d8ddb3a0fdc8fb027f10684210191e0619b71986/drivers/hmis/app/graphql/mutations/submit_form.rb#L17) we wrap the whole SubmitForm mutation in a transaction, but from within the mutation we call `enrollment.save_new_enrollment!` which, as mentioned, uses a nested transaction to delete the wip record and save the enrollment record.

I feel like the best way to deal with this is to remove transactions from the model helper methods. It is on the "caller" (eg the mutation code) to ensure that it is called within a transaction if it is updating multiple records. Does that make sense? Any other ideas?

LMK if you agree this is an issue, broadly. I will open a separate ticket.

Reply:
> From my understanding, we definitely DON'T want nested transactions within a mutation execution, because that would lead to rollbacks being ignored

Yes, it's also my understanding that raising `ActiveRecord::Rollback` within a nested transaction has unintuitive results. The rollback exception is special though and is used to fail the transaction without breaking execution flow. Looks like there's only 9 occurrences of this in our codebase currently. 

I believe other kinds of exceptions should bubble up to the outer transaction and prevent it from being committed. For example, if `save!` fails within a nested tx then the error should cause whole transaction to fail (unless the error is explicitly rescued within the tx).

So if we have cases where we are raising a rollback exception within a nested transaction, we should fix that. I don't see that we are doing that right now though although maybe I'm missing it? 

I agree that it's best to let the caller manage the transaction, so I'm in favor of pulling transactions up into the mutation. However I don't know that our current usage would cause problems.

Reply:
Neeevermind..  I tested this, and it's not what I thought. TLDR no cause for concern, and I will indeed add a transaction here, even though it will lead to a nested transaction.

If there is an exception raised, it will still break out of the parent transaction and roll back everything in the parent transaction. The only [exception](https://github.com/adjust/streets/blob/677224ccebbbab210b0c852f2824b0c60045d1e8/activerecord/lib/active_record/transactions.rb#L101) is ActiveRecord::Rollback error, which will actually not roll back the changes when in a nested block:

```
def test_transactions
  Hmis::Hud::Base.transaction do
    # outer change
    Hmis::Hud::Client.find(58726).update(last_name: "changed last name")

    Hmis::Hud::Base.transaction do
      # inner change
      Hmis::Hud::Client.find(58726).update(first_name: "changed first name")

      # raise "something went wrong"    # <==  both changes are rolled back
      raise ActiveRecord::Rollback # <==   neither change is rolled back
    end
  end
end

Hmis::Hud::Client.find(58726).update(first_name: "pom", last_name: "jones"); test_transactions; Hmis::Hud::Client.find(58726).full_name
```

There are only a handful of places in the driver where we raise a `raise ActiveRecord::Rollback`, and none of them look problematic. But this is good for me to know, I didn't understand this distinction before.

cc @martha for learning purposes 🦀

Reply:
Whoops somehow I didn't see your comment before responding @ttoomey! THank you for the explanation, I got there 😅  should've done my own testing first